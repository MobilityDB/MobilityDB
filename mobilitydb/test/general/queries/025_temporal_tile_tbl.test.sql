-------------------------------------------------------------------------------
--
-- This MobilityDB code is provided under The PostgreSQL License.
-- Copyright (c) 2016-2024, Universit√© libre de Bruxelles and MobilityDB
-- contributors
--
-- MobilityDB includes portions of PostGIS version 3 source code released
-- under the GNU General Public License (GPLv2 or later).
-- Copyright (c) 2001-2024, PostGIS contributors
--
-- Permission to use, copy, modify, and distribute this software and its
-- documentation for any purpose, without fee, and without a written
-- agreement is hereby granted, provided that the above copyright notice and
-- this paragraph and the following two paragraphs appear in all copies.
--
-- IN NO EVENT SHALL UNIVERSITE LIBRE DE BRUXELLES BE LIABLE TO ANY PARTY FOR
-- DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING
-- LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
-- EVEN IF UNIVERSITE LIBRE DE BRUXELLES HAS BEEN ADVISED OF THE POSSIBILITY
-- OF SUCH DAMAGE.
--
-- UNIVERSITE LIBRE DE BRUXELLES SPECIFICALLY DISCLAIMS ANY WARRANTIES,
-- INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
-- AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS ON
-- AN "AS IS" BASIS, AND UNIVERSITE LIBRE DE BRUXELLES HAS NO OBLIGATIONS TO
-- PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
--
-------------------------------------------------------------------------------

-------------------------------------------------------------------------------
-- Multidimensional tiling
-------------------------------------------------------------------------------

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(i, 2) AS bl FROM tbl_intspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(i, 2, 1) AS bl FROM tbl_intspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(f, 2) AS bl FROM tbl_floatspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(f, 2.5, 1.5) AS bl FROM tbl_floatspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(t, '2 days') AS bl FROM tbl_tstzspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (bl).index, COUNT((bl).span) FROM (SELECT bins(t, '2 days', '2001-06-01') AS bl FROM tbl_tstzspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

-------------------------------------------------------------------------------

SELECT getBin(i, 2), COUNT(*) FROM tbl_int GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT getBin(i, 2, 1), COUNT(*) FROM tbl_int GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

SELECT getBin(f, 2.5), COUNT(*) FROM tbl_float GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT getBin(f, 2.5, 1.5), COUNT(*) FROM tbl_float GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

SELECT getBin(t, interval '2 days'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT getBin(t, interval '2 days', timestamptz '2001-06-01'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

-------------------------------------------------------------------------------

SELECT valueTimeTiles(b, 2.5, '1 week'), COUNT(*) FROM tbl_tboxfloat GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT valueTimeTiles(b, 2.5, '1 week', 1.5), COUNT(*) FROM tbl_tboxfloat GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT valueTimeTiles(b, 2.5, '1 week', 1.5, '2001-06-01'), COUNT(*) FROM tbl_tboxfloat GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

-- Time
SELECT extent(getTboxTimeTile(t2.t, '1 week')) FROM
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;
SELECT extent(getTboxTimeTile(t2.t, '1 week', '2001-01-15')) FROM
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;
-- 2D
SELECT extent(getValueTimeTile(t1.f, t2.t, 2.5, '1 week')) FROM
(SELECT * FROM tbl_float WHERE f IS NOT NULL LIMIT 10) t1,
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;
SELECT extent(getValueTimeTile(t1.f, t2.t, 2.5, '1 week', 3.5, '2001-01-15')) FROM
(SELECT * FROM tbl_float WHERE f IS NOT NULL LIMIT 10) t1,
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;

-------------------------------------------------------------------------------
-- valueTimeBoxes
-------------------------------------------------------------------------------

SELECT SUM(array_length(valueTimeBoxes(temp, 2, '1 week'), 1)) FROM tbl_tint;
SELECT SUM(array_length(valueTimeBoxes(temp, 2, '1 week', 1), 1)) FROM tbl_tint;
SELECT SUM(array_length(valueTimeBoxes(temp, 2, '1 week', 1, '2001-06-01'), 1)) FROM tbl_tint;

SELECT SUM(array_length(valueTimeBoxes(temp, 2.5, '1 week'), 1)) FROM tbl_tfloat;
SELECT SUM(array_length(valueTimeBoxes(temp, 2.5, '1 week', 1.5), 1)) FROM tbl_tfloat;
SELECT SUM(array_length(valueTimeBoxes(temp, 2.5, '1 week', 1.5, '2001-06-01'), 1)) FROM tbl_tfloat;

-------------------------------------------------------------------------------
-- valueSplit
-------------------------------------------------------------------------------

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2) AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2, 1) AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2.5) AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2.5, 1.5) AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

WITH temp1 AS (
  SELECT k, temp, (tb).tnumber AS slice
  FROM (SELECT k, temp, valueSplit(temp, 5) AS tb from tbl_tfloat) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS tmerge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> tmerge ORDER BY k;

-------------------------------------------------------------------------------
-- timeSplit
-------------------------------------------------------------------------------

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tbool) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tbool) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_ttext) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_ttext) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

WITH temp1 AS (
  SELECT k, temp, (tb).temp AS slice
  FROM (SELECT k, temp, timeSplit(temp, '5 min') AS tb FROM tbl_tfloat) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS tmerge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> tmerge ORDER BY k;

-------------------------------------------------------------------------------
-- valueTimeSplit
-------------------------------------------------------------------------------

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2, '2 days') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2, '2 days', 1, '2001-06-01') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2.5, '2 days') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2.5, '2 days', 1.5, '2001-06-01') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;

WITH temp1 AS (
  SELECT k, temp, (tb).tnumber AS slice
  FROM (SELECT k, temp, valueTimeSplit(temp, 5, '5 min') AS tb FROM tbl_tfloat) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS tmerge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> tmerge ORDER BY k;

-------------------------------------------------------------------------------
