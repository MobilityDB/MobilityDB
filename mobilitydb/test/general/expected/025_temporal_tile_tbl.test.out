SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(i, 2) AS bl FROM tbl_intspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
     2 |    88
     3 |    66
(3 rows)

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(i, 2, 1) AS bl FROM tbl_intspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
     2 |    84
     3 |    65
(3 rows)

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(f, 2) AS bl FROM tbl_floatspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
     2 |    94
     3 |    75
(3 rows)

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(f, 2.5, 1.5) AS bl FROM tbl_floatspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
     2 |    92
     3 |    63
(3 rows)

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(p, '2 days') AS bl FROM tbl_tstzspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
(1 row)

SELECT (bl).index, COUNT((bl).span) FROM (SELECT bucketList(p, '2 days', '2001-06-01') AS bl FROM tbl_tstzspan) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 index | count 
-------+-------
     1 |    99
(1 row)

SELECT SUM(valueBucket(i, 2)) FROM tbl_int;
 sum  
------
 4828
(1 row)

SELECT SUM(valueBucket(i, 2, 1)) FROM tbl_int;
 sum  
------
 4825
(1 row)

SELECT SUM(valueBucket(f, 2.5)) FROM tbl_float;
  sum   
--------
 4712.5
(1 row)

SELECT SUM(valueBucket(f, 2.5, 1.5)) FROM tbl_float;
 sum  
------
 4706
(1 row)

SELECT spanBucket(i, 2), COUNT(*) FROM tbl_int GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 spanbucket | count 
------------+-------
 [74, 76)   |     6
 [2, 4)     |     5
 [24, 26)   |     5
(3 rows)

SELECT spanBucket(i, 2, 1), COUNT(*) FROM tbl_int GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 spanbucket | count 
------------+-------
 [73, 75)   |     6
 [1, 3)     |     5
 [21, 23)   |     5
(3 rows)

SELECT spanBucket(f, 2.5), COUNT(*) FROM tbl_float GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 spanbucket | count 
------------+-------
 [15, 17.5) |     6
 [45, 47.5) |     6
 [10, 12.5) |     5
(3 rows)

SELECT spanBucket(f, 2.5, 1.5), COUNT(*) FROM tbl_float GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 spanbucket | count 
------------+-------
 [14, 16.5) |     7
 [9, 11.5)  |     6
 [46.5, 49) |     6
(3 rows)

SELECT timeBucket(t, '1 week'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
          timebucket          | count 
------------------------------+-------
 Mon Jun 11 01:00:00 2001 PDT |     6
 Mon Sep 17 01:00:00 2001 PDT |     6
 Mon Jan 22 00:00:00 2001 PST |     5
(3 rows)

SELECT timeBucket(t, '1 week', timestamptz '2001-06-01'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
          timebucket          | count 
------------------------------+-------
 Fri Sep 14 00:00:00 2001 PDT |     8
 Thu Mar 29 23:00:00 2001 PST |     5
 Fri Jun 15 00:00:00 2001 PDT |     5
(3 rows)

SELECT periodBucket(t, interval '2 days'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
                         periodbucket                         | count 
--------------------------------------------------------------+-------
 [Sun Sep 16 01:00:00 2001 PDT, Tue Sep 18 01:00:00 2001 PDT) |     4
 [Sat Jan 27 00:00:00 2001 PST, Mon Jan 29 00:00:00 2001 PST) |     3
 [Sat Oct 06 01:00:00 2001 PDT, Mon Oct 08 01:00:00 2001 PDT) |     3
(3 rows)

SELECT periodBucket(t, interval '2 days', timestamptz '2001-06-01'), COUNT(*) FROM tbl_timestamptz GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
                         periodbucket                         | count 
--------------------------------------------------------------+-------
 [Mon Sep 17 00:00:00 2001 PDT, Wed Sep 19 00:00:00 2001 PDT) |     4
 [Wed Apr 18 00:00:00 2001 PDT, Fri Apr 20 00:00:00 2001 PDT) |     3
 [Thu May 10 00:00:00 2001 PDT, Sat May 12 00:00:00 2001 PDT) |     3
(3 rows)

SELECT tileList(b, 2.5, '1 week'), COUNT(*) FROM tbl_tbox GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
                                        tilelist                                        | count 
----------------------------------------------------------------------------------------+-------
 (1,"TBOX XT([10, 12.5),[Mon Jan 08 00:00:00 2001 PST, Mon Jan 15 00:00:00 2001 PST))") |     2
 (1,"TBOX XT([5, 7.5),[Mon May 28 01:00:00 2001 PDT, Mon Jun 04 01:00:00 2001 PDT))")   |     2
 (1,"TBOX XT([35, 37.5),[Mon Nov 19 00:00:00 2001 PST, Mon Nov 26 00:00:00 2001 PST))") |     2
(3 rows)

SELECT tileList(b, 2.5, '1 week', 1.5), COUNT(*) FROM tbl_tbox GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
                                        tilelist                                        | count 
----------------------------------------------------------------------------------------+-------
 (1,"TBOX XT([34, 36.5),[Mon Nov 19 00:00:00 2001 PST, Mon Nov 26 00:00:00 2001 PST))") |     2
 (1,"TBOX XT([6.5, 9),[Mon Dec 03 00:00:00 2001 PST, Mon Dec 10 00:00:00 2001 PST))")   |     2
 (2,"TBOX XT([36.5, 39),[Mon Nov 19 00:00:00 2001 PST, Mon Nov 26 00:00:00 2001 PST))") |     2
(3 rows)

SELECT tileList(b, 2.5, '1 week', 1.5, '2001-06-01'), COUNT(*) FROM tbl_tbox GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
                                        tilelist                                        | count 
----------------------------------------------------------------------------------------+-------
 (1,"TBOX XT([11.5, 14),[Thu Jan 11 23:00:00 2001 PST, Thu Jan 18 23:00:00 2001 PST))") |     2
 (1,"TBOX XT([11.5, 14),[Thu Feb 22 23:00:00 2001 PST, Thu Mar 01 23:00:00 2001 PST))") |     2
 (1,"TBOX XT([34, 36.5),[Thu Nov 22 23:00:00 2001 PST, Thu Nov 29 23:00:00 2001 PST))") |     2
(3 rows)

SELECT extent(tile(t1.f, t2.t, 2.5, '1 week')) FROM
(SELECT * FROM tbl_float WHERE f IS NOT NULL LIMIT 10) t1,
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;
                                    extent                                     
-------------------------------------------------------------------------------
 TBOX XT([5, 85),[Mon Jan 22 00:00:00 2001 PST, Mon Dec 10 00:00:00 2001 PST))
(1 row)

SELECT extent(tile(t1.f, t2.t, 2.5, '1 week', 3.5, '2001-01-15')) FROM
(SELECT * FROM tbl_float WHERE f IS NOT NULL LIMIT 10) t1,
(SELECT * FROM tbl_timestamptz WHERE t IS NOT NULL LIMIT 10) t2;
                                     extent                                      
---------------------------------------------------------------------------------
 TBOX XT([6, 83.5),[Mon Jan 22 00:00:00 2001 PST, Mon Dec 10 00:00:00 2001 PST))
(1 row)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2) AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
     48 |    20
     64 |    18
     44 |    17
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2, 1) AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
     25 |    19
     41 |    18
     49 |    18
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2.5) AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
   67.5 |    50
     55 |    48
     65 |    48
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueSplit(temp, 2.5, 1.5) AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
   66.5 |    48
     69 |    48
     24 |    47
(3 rows)

WITH temp1 AS (
  SELECT k, temp, (tb).tnumber AS slice
  FROM (SELECT k, temp, valueSplit(temp, 5) AS tb from tbl_tfloat_big) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS merge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> merge ORDER BY k;
 k 
---
(0 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tbool) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Tue Feb 06 14:00:00 2001 PST |     2
 Sat Feb 17 08:00:00 2001 PST |     2
 Wed Feb 28 10:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tbool) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Tue Feb 06 13:00:00 2001 PST |     2
 Sat Feb 17 09:00:00 2001 PST |     2
 Wed Feb 28 09:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Fri Jun 29 11:00:00 2001 PDT |     3
 Sun Feb 04 02:00:00 2001 PST |     2
 Sun Feb 04 04:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Fri Jun 29 12:00:00 2001 PDT |     3
 Sun Feb 04 03:00:00 2001 PST |     2
 Sun Feb 11 23:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Fri Jan 12 04:00:00 2001 PST |     2
 Fri Jan 26 18:00:00 2001 PST |     2
 Tue Jan 30 14:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Fri Jan 12 03:00:00 2001 PST |     2
 Fri Jan 26 17:00:00 2001 PST |     2
 Tue Jan 30 15:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours') AS sp FROM tbl_ttext) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Sun Dec 02 00:00:00 2001 PST |     3
 Sun Dec 02 02:00:00 2001 PST |     3
 Fri Jan 05 06:00:00 2001 PST |     2
(3 rows)

SELECT (sp).time, COUNT((sp).temp) FROM (SELECT timeSplit(temp, '2 hours', '2001-06-01') AS sp FROM tbl_ttext) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
             time             | count 
------------------------------+-------
 Sat Dec 01 23:00:00 2001 PST |     3
 Fri Jan 05 05:00:00 2001 PST |     2
 Fri Jan 05 07:00:00 2001 PST |     2
(3 rows)

WITH temp1 AS (
  SELECT k, temp, (tb).temp AS slice
  FROM (SELECT k, temp, timeSplit(temp, '5 min') AS tb FROM tbl_tfloat_big) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS merge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> merge ORDER BY k;
 k 
---
(0 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2, '2 days') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
     48 |    20
     20 |    18
     64 |    18
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2, '2 days', 1, '2001-06-01') AS sp FROM tbl_tint) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
     25 |    19
     41 |    18
     49 |    18
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2.5, '2 days') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
   67.5 |    52
     55 |    50
     65 |    50
(3 rows)

SELECT (sp).number, COUNT((sp).tnumber) FROM (SELECT valueTimeSplit(temp, 2.5, '2 days', 1.5, '2001-06-01') AS sp FROM tbl_tfloat) t GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 3;
 number | count 
--------+-------
   66.5 |    48
     69 |    48
     24 |    47
(3 rows)

WITH temp1 AS (
  SELECT k, temp, (tb).tnumber AS slice
  FROM (SELECT k, temp, valueTimeSplit(temp, 5, '5 min') AS tb FROM tbl_tfloat_big) t ),
temp2 AS (
  SELECT k, temp, merge(slice ORDER BY slice) AS merge
  FROM temp1 GROUP BY k, temp )
SELECT k FROM temp2 WHERE temp <> merge ORDER BY k;
 k 
---
(0 rows)

